cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/templates/etl/batch/ETLBatchTemplate.java

```diff
@@ -100,6 +103,49 @@ public void configureAdapter(String adapterName, ETLBatchConfig etlConfig, Adapt
                                             cronEntry));
   }
 
+  private void validateAdapter(ETLStage source, ETLStage sink, List<ETLStage> transform) throws Exception {
+    String sourceName = source.getName();
+    String sinkName = sink.getName();
+    String sourceClassName = sourceClassMap.get(sourceName);
+    String sinkClassName = sinkClassMap.get(sinkName);
+
+    BatchSource batchSource = (BatchSource) Class.forName(sourceClassName).newInstance();
+    BatchSink batchSink = (BatchSink) Class.forName(sinkClassName).newInstance();
+
+    if (transform.size() == 0) {
+      // No transforms. Check only source and sink.
+      Preconditions.checkArgument(batchSink.getKeyType().equals(batchSource.getKeyType()));
```

Comment:
should give a meaningful error message, otherwise people will just be confused by the exception.